internal/venice-common/src/main/java/com/linkedin/venice/pubsub/adapter/PubSubSharedProducerFactory.java

```diff
@@ -0,0 +1,210 @@
+          leastLoadedSharedProducer = producers[i];
+        }
+        continue;
+      }
+      LOGGER.info("Creating shared producer with id: {}", i);
+      sharedProducerAdapter = createSharedProducer(i);
+      producers[i] = sharedProducerAdapter;
+      LOGGER.info("Created shared producer instance: {}", sharedProducerAdapter);
+      incrActiveSharedProducerCount();
+      break;
+    }
+
+    // Use the least used producer instance
+    if (sharedProducerAdapter == null) {
+      if (leastLoadedSharedProducer == null) {
+        throw new VeniceException("No shared producer available");
+      }
+      sharedProducerAdapter = leastLoadedSharedProducer;
+    }
+
+    sharedProducerAdapter.addProducerTask(producerTaskName);
+    producerTaskToProducerMap.put(producerTaskName, sharedProducerAdapter);
+    LOGGER.info("Acquired the shared producer: {} for:{}", sharedProducerAdapter.getId(), producerTaskName);
+    incrActiveSharedProducerTasksCount();
+    return sharedProducerAdapter;
+  }
+
+  public synchronized void releaseSharedProducer(String producerTaskName) {
+    if (!isRunning) {
+      throw new VeniceException(
+          "Shared producer factory has been already closed, can't release the producer for task:" + producerTaskName);
+    }
+
+    if (!producerTaskToProducerMap.containsKey(producerTaskName)) {
+      LOGGER.error("Shared producer factory does not have a producer for:{}", producerTaskName);
+      return;
+    }
+    PubSubSharedProducerAdapter sharedProducerAdapter = producerTaskToProducerMap.get(producerTaskName);
+    sharedProducerAdapter.removeProducerTask(producerTaskName);
+    producerTaskToProducerMap.remove(producerTaskName, sharedProducerAdapter);
+    LOGGER.info("{} released the producer id: {}", producerTaskName, sharedProducerAdapter.getId());
+    decrActiveSharedProducerTasksCount();
+  }
+
+  /**
+   * N.B. This shared producer does not support brokerAddressToOverride. Do not use this in controllers.
+   *
+   * @param veniceProperties
+   * @return An instance of shared producer adapter
+   */
+  @Override
+  public PubSubSharedProducerAdapter create(
+      VeniceProperties veniceProperties,
+      String topicName,
+      String brokerAddressToOverride) {
+    return acquireSharedProducer(topicName);
+  }
+
+  /**
+   * @param id Used as unique identifier to construct a new shared producer
+   * @return An instance of shared producer adapter
+   */
+  public abstract PubSubSharedProducerAdapter createSharedProducer(int id);
+
+  /**
+   * @return returns the name of shared producer factory
+   */
+  public abstract String getName();
+
+  public long getActiveSharedProducerTasksCount() {
+    return activeSharedProducerTasksCount.get();
+  }
+
+  public long getActiveSharedProducerCount() {
+    return activeSharedProducerCount.get();
+  }
+
+  private void incrActiveSharedProducerTasksCount() {
+    activeSharedProducerTasksCount.incrementAndGet();
```

Comment:
nit: since it is void return just increment() should be fine.

Reply:
IMO there is no `increment()` method. We can use either `getAndIncrement` or `incrementAndGet`